include/range/v3/detail/config.hpp

```diff
@@ -22,15 +22,42 @@
 #define RANGES_ASSERT assert
 #endif
 
+#if defined(__GNUC__) || defined(__clang__)
+#define RANGES_ASSUME(COND) ((COND) ? void() : __builtin_unreachable())
+#elif defined(_MSC_VER)
+#define RANGES_ASSUME(COND) __assume(COND)
+#endif
+
+#ifndef RANGES_CONTRACT_VIOLATIONS_THROW
+#define RANGES_CONTRACT_VIOLATIONS_THROW 0
+#endif
+
 #ifndef RANGES_ENSURE_MSG
-#include <exception>
-#define RANGES_ENSURE_MSG(COND, MSG) \
-    ((COND) ? void() : (RANGES_ASSERT(!(true && MSG)), std::terminate()))
+#if RANGES_CONTRACT_VIOLATIONS_THROW
+#include <stdexcept>
+namespace ranges
+{
+    inline namespace v3
+    {
+        struct contract_violation
```

Comment:
I need to think about this. It's not How Things Are Done (tm) in range-v3, and the nastiness you refer to is indeed nasty. I gave people the option to #define `RANGES_ASSERT` in the hopes that they could use that as a hook for contract violations, but making it throw doesn't work because of the `noexcept` business. Let me noodle.